packages/aws-cdk-lib/aws-s3-deployment/lib/bucket-deployment.ts

```diff
@@ -377,9 +377,9 @@ export class BucketDeployment extends Construct {
       serviceToken: handler.functionArn,
       resourceType: 'Custom::CDKBucketDeployment',
       properties: {
-        SourceBucketNames: cdk.Lazy.list({ produce: () => this.sources.map(source => source.bucket.bucketName) }),
-        SourceObjectKeys: cdk.Lazy.list({ produce: () => this.sources.map(source => source.zipObjectKey) }),
-        SourceMarkers: cdk.Lazy.any({
+        SourceBucketNames: cdk.Lazy.uncachedList({ produce: () => this.sources.map(source => source.bucket.bucketName) }),
+        SourceObjectKeys: cdk.Lazy.uncachedList({ produce: () => this.sources.map(source => source.zipObjectKey) }),
+        SourceMarkers: cdk.Lazy.uncachedAny({
```

Comment:
I'm sure this is needed.... but why?

Reply:
Good question.

TL;DR - this is needed to ensure the nested stack template file is added as a `Source` to the `BucketDeployment`.

Take this code block as an example:

```ts
class SampleNestedStack extends cdk.NestedStack {
      constructor(scope: Construct, id: string) {
        super(scope, id);

        new lambda.Function(this, 'HelloHandler1', {
          runtime: lambda.Runtime.PYTHON_3_9,
          code: lambda.Code.fromAsset(path.join(__dirname, 'assets')),
          handler: 'index.handler',
        });
        
        new lambda.Function(this, 'HelloHandler2', {
          runtime: lambda.Runtime.PYTHON_3_9,
          code: lambda.Code.fromAsset(path.join(__dirname, 'assetsv2')),
          handler: 'index.handler',
        });
      }
    }
```

Note: I'm not a `Lazy` expert, but the process is something like this. I had some help from @rix0rrr to figure this part out.

First, the `Code` assets from the two lambda `Functions` will be added as `Sources` to the `BucketDeployment`. Since these are evaluated with `Lazy` they are added to a cache before evaluation is carried out. At the point when the `NestedStack` is ready to add its own template, the `Lazy` logic has already assumed all assets have been added to the cache. This results in the template file for `NestedStack` being left out when `SourceBucketNames`, `SourceObjectKeys`, and `SourceMarkers` is evaluated. `Lazy.uncachedList` is basically saying don't only rely on the cache, there may be more items to evaluate.

Reply:
Btw, probably not necessary for the markers. Doesn’t really harm, but maybe put it back? And add a comment saying that its uncached because the contents may change for nestedstacks in a productstacj